doc/neps/nep-XXXX-SIMD-optimizations.rst

```diff
@@ -0,0 +1,229 @@
+To aid with this, the implementation for this NEP will add a way to select
+which instruction sets can be used at *runtime* via environment variables.
+(name TBD).
+
+
+Diagnostics
+```````````
+
+A new dictionary `__cpu_features__` will be available to python. The keys are
+the available features, the value is a boolean ``True``. Various new private
+C functions will be used internally to query available features. These
+might be exposed via specific c-extension modules for testing.
+
+
+Workflow for adding a new CPU architecture-specific optimization
+````````````````````````````````````````````````````````````````
+
+NumPy will always have a baseline C implementation for any code that may be
+a candidate for SIMD vectorization.  If a contributor wants to add SIMD
+support for some architecture (typically the one of most interest to them),
+this is the proposed workflow:
+
+TODO (see https://github.com/numpy/numpy/pull/13516#issuecomment-558859638,
+needs to be worked out more)
+
+Reuse by other projects
+```````````````````````
+
+It would be nice if the universal intrinsics would be available to other
+libraries like SciPy or Astropy that also build ufuncs, but that is not an
+explicit goal of the first implementation of this NEP.
+
+Backward compatibility
+----------------------
+
+There should be no impact on backwards compatibility.
+
+
+Detailed description
+--------------------
+
+*This section should provide a detailed description of the proposed change.
+It should include examples of how the new functionality would be used,
+intended use-cases and pseudo-code illustrating its use.*
+
+TODO: status today - what instructions are used at build time (SSE2/3 and
+AVX/AVX512 for XXX functionality) and at runtime (some but less, see
+``loops.c.src``)
+
+
+Related Work
+------------
+
+- PIXMAX TBD: what is it?
+- `Eigen`_ is a C++ template library for linear algebra: matrices, vectors,
+  numerical solvers, and related algorithms. It is a higher level-abstraction
+  than the intrinsics discussed here.
+- `xsimd`_ is a header-only C++ library for x86 and ARM that implements the
+  mathematical functions used in the algorithms of ``boost.SIMD``.
+- OpenCV used to have the one-implementation-per-architecture design, but more
+  recently moved to a design that is quite similar to what is proposed in this
+  NEP. The top-level `dispatch code`_ includes a `generic header`_ that is
+  `specialized at compile time`_ by the CMakefile system.
+
+
+Implementation
+--------------
+
+Current PRs:
+
+- `gh-13421 improve runtime detection of CPU features <https://github.com/numpy/numpy/pull/13421>`_
+- `gh-13516: enable multi-platform SIMD compiler optimizations <https://github.com/numpy/numpy/pull/13516>`_
+
+**Let's leave description of this out for now. Only do that once the questions
+in the sections above are answered.**
+
+
+Alternatives
+------------
+
+A proposed alternative in gh-13516_ is a per CPU architecture implementation of
```

Comment:
The alternative solution can be achieved within the current design of infrastructure(gh-13516), by treating dispatch-able sources as a router.
For example :
```C
/*@targets baseline SSE2 AVX2 AVX512F NEON VSX  */
#include "npy_cpu_dispatch.h"
// add some uitlities and headers
#if defined(NPY_HAVE_AVX512F)
    #include "loops.avx512.c"
#elif defined(NPY_HAVE_AVX2)
    #include "loops.avx2.c"
#elif defined(NPY_HAVE_SSE2)
    #include "loops.sse2.c"
#elif defined(NPY_HAVE_VSX)
    #include "loops.vsx.c"
#elif defined(NPY_HAVE_NEON)
    #include "loops.neon.c"
#else
    // write C scalars code
#endif
````

Reply:
The intent here was more along the lines of "each platform has a champion who adds platform-specific code without trying to abstract common intrinsics". I will try to reformulate.